python/plugins/processing/algs/qgis/RasterSampling.py

```diff
@@ -0,0 +1,163 @@
+*                                                                         *
+***************************************************************************
+"""
+
+__author__ = 'Matteo Ghetta'
+__date__ = 'July 2018'
+__copyright__ = '(C) 2018, Matteo Ghetta'
+
+# This will get replaced with a git SHA1 when you do a git archive
+
+__revision__ = '$Format:%H$'
+
+
+import os
+
+from qgis.PyQt.QtGui import QIcon
+from qgis.PyQt.QtCore import QVariant
+
+from qgis.core import (QgsApplication,
+                       QgsField,
+                       QgsFeatureSink,
+                       QgsRaster,
+                       QgsProcessing,
+                       QgsProcessingParameterMultipleLayers,
+                       QgsFields,
+                       QgsProcessingUtils,
+                       QgsProcessingException,
+                       QgsProcessingParameterFeatureSource,
+                       QgsProcessingParameterFeatureSink)
+
+from processing.algs.qgis.QgisAlgorithm import QgisAlgorithm
+
+
+class RasterSampling(QgisAlgorithm):
+
+    INPUT = 'INPUT'
+    RASTERCOPY = 'RASTERCOPY'
+    OUTPUT = 'OUTPUT'
+
+    def name(self):
+        return 'rastersampling'
+
+    def displayName(self):
+        return self.tr('Sample Raster Values')
+
+    def group(self):
+        return self.tr('Raster analysis')
+
+    def groupId(self):
+        return 'rasteranalysis'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(
+            QgsProcessingParameterFeatureSource(
+                self.INPUT,
+                self.tr('Input Point Layer'),
+                [QgsProcessing.TypeVectorPoint]
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterMultipleLayers(
+                self.RASTERCOPY,
+                self.tr('Raster Layer to sample'),
+                QgsProcessing.TypeRaster
+            )
+        )
+
+        self.addParameter(
+            QgsProcessingParameterFeatureSink(
+                self.OUTPUT,
+                self.tr('Sampled Points')
+            )
+        )
+
+    def processAlgorithm(self, parameters, context, feedback):
+
+        source = self.parameterAsSource(
+            parameters,
+            self.INPUT,
+            context
+        )
+
+        sampled_rasters = self.parameterAsLayerList(
+            parameters,
+            self.RASTERCOPY,
+            context
+        )
+
+        if source is None:
+            raise QgsProcessingException(self.invalidSourceError(parameters, self.INPUT))
+
+        source_fields = source.fields()
+        raster_fields = QgsFields()
+
+        # append field to vector as rasterName_bandCount
+        for i in sampled_rasters:
+            for b in range(i.bandCount()):
+                raster_fields.append(QgsField(
+                    os.path.splitext(os.path.basename(i.source()))[0] + str('_{}'.format(b + 1)), QVariant.Double)
```

Comment:
This is not safe for non-gdal providers. E.g. for a wms layer the source isn't a local file, so os.path.basename will return... something else.

But see comment below regarding this!